zcash_client_sqlite/src/lib.rs

```diff
@@ -413,32 +418,24 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                         wallet::sapling::put_received_note(&wdb.conn.0, output, tx_row)?;
 
                     // Save witness for note.
-                    new_witnesses.push((received_note_id, output.witness().clone()));
+                    wallet_note_ids.push(received_note_id);
                 }
             }
 
-            // Insert current new_witnesses into the database.
-            for (received_note_id, witness) in updated_witnesses.iter().chain(new_witnesses.iter())
-            {
-                if let NoteId::ReceivedNoteId(rnid) = *received_note_id {
-                    wallet::sapling::insert_witness(
-                        &wdb.conn.0,
-                        rnid,
-                        witness,
-                        block.block_height,
-                    )?;
-                } else {
-                    return Err(SqliteClientError::InvalidNoteId);
+            let mut sapling_commitments = block.sapling_commitments.into_iter();
+            wdb.with_sapling_tree_mut::<_, _, SqliteClientError>(move |sapling_tree| {
+                if let Some(sapling_tree_size) = block.sapling_commitment_tree_size {
+                    let start_position = Position::from(u64::from(sapling_tree_size))
+                        - u64::try_from(sapling_commitments.len()).unwrap();
+                    sapling_tree.batch_insert(start_position, &mut sapling_commitments)?;
```

Comment:
```
error[E0507]: cannot move out of `sapling_commitments`, a captured variable in an `FnMut` closure
   --> zcash_client_sqlite/src/lib.rs:432:63
    |
427 |             let sapling_commitments = block.sapling_commitments;
    |                 ------------------- captured outer variable
428 |             wdb.with_sapling_tree_mut::<_, _, SqliteClientError>(move |sapling_tree| {
    |                                                                  ------------------- captured by this `FnMut` closure
...
432 |                     sapling_tree.batch_insert(start_position, sapling_commitments.into_iter())?;
    |                                                               ^^^^^^^^^^^^^^^^^^^ ----------- `sapling_commitments` moved due to this method call
    |                                                               |
    |                                                               move occurs because `sapling_commitments` has type `Vec<(zcash_primitives::sapling::Node, Retention<BlockHeight>)>`, which does not implement the `Copy` trait
```